### DISCUSSION TITLE
Dockerhub build and Mathics 5.0.0 release thoughts

### DISCUSSION THREAD

**Discussion by rocky** [Mathics3/mathics-core#438] [General]
Docker update from last week
----------------------------------------

Last weekend I updated the [Mathics docker image on dockerhub](https://hub.docker.com/repository/docker/mathicsorg/mathics) to work with the current development 5.0.0x sources as of last week.  

It was a heavy lift since in that image we switched the docker build base from Ubuntu 20.04 to Ubuntu 22.4. Ubuntu 22.04 uses Python 3.10 as its system Python. For Django  though we use pyston-2.3.4 (based on Python 3.8) since that is much faster. Pyston doesn't handle matplotlib rendering used in mathicsscript, so for mathicsscript we have to use CPython.

In the process of building the docker image, I learned that a lot was broken that I wasn't aware of previously. In particular, all of the Pymathics modules like the ones for NLP, and Graph Networks were broken since the Mathics API has changed.

In the last docker build I didn't upgrade Asymptote from 2.78 to the current 2.81 because I noticed a long-standing intermittent problem in showing 2D Graphics "filling" options (with values Top, Bottom, and Axis). I didn't have time to investigate that or the other problems that weekend. 

However the asymptote bug  preventing the filling option from working in PDFs has been fixed in commit [18a62ea6835994d439c30381cc458473f4767fb7](https://github.com/Mathics3/mathics-core/commit/312ae1aae0b55872234b3d27af1230bf6ace50c2). The bug had nothing to do with the version of asymptote. (I thought it might have something to do with the aymptote version because previously, there was a bug in asymptote 2.79 that prevented us from using that.)

Docker update this weekend
--------------------------------------

This weekend I plan on putting out yet another docker update for the changes that have happened since last week which has been more or less all the bug fixes to bugs noticed in making the docker build.  So here we will now be able to use a more recent version of Asymptote, and the NLP and Networkx Graph packages will be reinstated.  Also since last week I have largely been working on those long-term organizational things where we create more sections that follow the Mathematica organization better rather than have the rough mishmash that has been around for a while. The doc organization still won't be fully done, that will probably take a number of further iterations. But I expect things to get better.  I have code to sort chapters alphabetically, but in day-to-day testing that causes huge slowdowns. But for more-or-less one-time PDF building that slowdown is okay to have. 

5.0.0 Release one of the subsequent weekends
-------------------------------------------------------------------

If I am successful this weekend, then more or less we are ready for a long overdue 5.0.0 release. In this release we start to clean up the Python Object classes around M-expressions and expression evaluation. 

In the release we start to make a small dent on the performance problems that have existed since the very first days.  There are tons of other things that would be great to do, and as mentioned in another discussion, Boxing and Formatting will be first on the list, but _after_ the release.

I expect the performance to improve greatly as we do that. And the long tail in expression evaluation improvement which started in 5.0.0 will continue.  I fully expect by the next release 2D and 3D plotting will be fixed. Probably the low-hanging fruit of better unicode use on output of expressions will be done then as well. I'd like to add more kinds of Plots like [`DiscretePlot`](https://reference.wolfram.com/language/ref/DiscretePlot.html) but that ties into the Boxing and Formatting as well, so I didn't want to start this release.

**Comment by rocky** [comment]
A Dockerhub Mathics happened last weekend and pymathics  packages were added, along with the newer asymptote and misc bug fixes noticed in preparing for a general release.

There are still some bugs in the pymathics modules and getting them to doctest and that I worked on Sunday. 

Basically the things I have been focusing are those things that it would be nice to fix or improve before a release. For example documentation cleanup is one of the things I typically do before a release and a bit of that has been done. (Of course, there is always more to do though).

If the pymathics stuff can be fixed without too much trouble, I'd like to see that go in and it would be a progression over the last 4.0.0 release from a year ago.

If it is too much work, it can stay as it is. 

Are there any other bugs that can easily be fixed?

**Comment by rocky** [comment]
Mathics, mathicsscript, mathics-django, pymathics-natlang 5.0.0  are now released and on github releases and PyPi.

I notice that the dice examples in the doc don't render, and there are some small things like this in the documentation, but overall I think things are a big win. 

@mmatera feel free to start working on Boxing and formatting. Sorry for the delay there.

**Reply by mmatera** [reply]
@rocky, thank you very much for all this work. Regarding the refactor of Boxing stuff, #482 contains the basic proposal.

**Reply by rocky** [reply]
> @rocky, thank you very much for all this work. Regarding the refactor of Boxing stuff, #482 contains the basic proposal

@mmatera  It looks very interesting and is much much needed. I would like to pull it down and try it, but it may take me a few days before I am able to.
